### DISCUSSION TITLE
Is extending signals endorsed?

### DISCUSSION THREAD

**Discussion by manuel-plavsic** [nank1ro/solidart#48] [General]
I am wondering what is the best way to manage complex signals, where a custom implementation to the `value` setter is wished. With Riverpod I would extend `StateNotifierProvider`. When using this library, is the following endorsed, or if there is a better approach?

```dart
class ThemeModeSignal<ThemeMode> extends Signal<ThemeMode> {
  ThemeModeSignal(super.initialValue);

  @override
  void set(ThemeMode newValue) {
    sharedPreferences.setString('themeMode', newValue.toString()); // for when app is restarted
    super.set(newValue);
  }
}
```

**Comment by nank1ro** [comment]
The code you provided is fine, the only downside I see is you can't extend an additional class (but you don't need to).
I am used to handling such cases in this way:
```dart
final themeMode = createSignal(ThemeMode.system);

themeMode.observe((previousValue, value) {
  sharedPreferences.setString('themeMode', value.toString());
});
```
But this is just a personal choice.

In any case, yes you can extend signals without any problems.
I think it should work fine also when using the `Solid` widget.

**Comment by manuel-plavsic** [comment]
That makes sense. Is `createEffect` also fine or does `ReadSignal.observe` have more benefits  (other than the previous value)?

**Reply by nank1ro** [reply]
`observe` and `createEffect` are slightly different, although in fact this code
```dart
counter.observe((previousValue, value) {
  print("$previousValue -> $value");
}, fireImmediately: true);
```

is equivalent to:
```dart
createEffect((_) {
  print("${counter.previousValue} -> ${counter.value}");
});
```

But `observe` is just for 1 signal, while `createEffect` can work with many signals, e.g.:
```dart
createEffect((_) {
  if (count() >=0 && doubleCount() <= 10) {
    // do something
  }
});
```
This logic will become pretty complex using `observe`.

Additionally, `observe` by default runs from the next value, unless you set `fireImmediately` to true.
Instead `createEffect` is always executed immediately

**Comment by manuel-plavsic** [comment]
How would you handle cases where `super.set` depends a lot on the `newValue`, such as this one:

```dart
class ThemeModeSignal<ThemeMode> extends Signal<ThemeMode> {
  ThemeModeSignal(super.initialValue);

  @override
  void set(ThemeMode newValue) {
    if (newValue == ThemeMode.system) {
      // ignore
      return;
    } else if (newValue == ThemeMode.dark) {
      super.set(ThemeMode.light);
    } else { // must be ThemeMode.light
      super.set(newValue);
    }
  }
}

**Reply by nank1ro** [reply]
It seems over complicated.
You may check how I implemented the toggle theme feature in Solidart [here](https://docs.page/nank1ro/solidart~dev/examples/toggle-theme)

If you don't care about the `ThemeMode.system` value, you may use a sub enum, like:
```dart
enum MyThemeMode {
  light,
  dark,
}
```

or even better, just use a boolean signal:
```dart
final isDarkMode = createSignal(true);
```

Additionally, why if the user wants the dark mode, you set the opposite value? It sounds very strange to me.

**Reply by manuel-plavsic** [reply]
The example I provided, where I set the opposite `ThemeData`, does not make sense. In hindsight, I should have added a note to clarify. However, there may be certain use cases where it is necessary to skip calling `set(newValue)`, or perhaps modify `newValue` before making the call. For instance, consider a scenario where you have a signal for setting design languages such as material, cupertino, macos_ui, fluent, and so on. In this case, it would be preferable to prevent an iOS/Android user from selecting the design language as macos_ui, as some parts of the screen may not be visible on a smaller screen.

However, upon further reflection, I realized that it would probably be better to handle this logic within the "UI parts" of the app. For example, if the device is an Android, then the macos_ui option should be completely disabled. This way, only `Signal` and `createEffect`/`observe` are needed.

When I started programming in Flutter, I heavily relied on `StateNotifierProvider`. However, this discussion has opened up a new perspective for me. When using Riverpod, I now believe that I should have utilized `ref.listen`/`ref.watch` more frequently in cases like the one mentioned above, instead of relying on `StateNotifierProvider`. It's always good to be open to learning better approaches. Thanks a lot for your tips!

**Reply by nank1ro** [reply]
You may skip the value before setting it:
```dart
if (newValue != ThemeMode.system) {
  themeMode.set(newValue);
}
```
What you say makes sense, it's more of an architectural talk than a state management talk.
I am of the opinion that the right way is the easiest way, consequently also the most maintainable way.
